migration/run.ts

```diff
@@ -0,0 +1,99 @@
+const base = `app`;
+const project = new Project({
+  useInMemoryFileSystem: true,
+  manipulationSettings: {
+    quoteKind: QuoteKind.Single
+  }
+});
+
+glob(`${base}/**/*.ts`, {}, (_, files) => {
+  const removeOnDestroy = process.argv.includes('--removeOnDestroy');
+
+  files.forEach(path => {
+    fs.readFile(path, 'utf8', (_, text) => {
+      if (!hasUntilDestroy.test(text)) return;
+
+      const result = transformCode(text, path, removeOnDestroy);
+
+      fs.writeFile(path, result, 'utf8', err => {
+        console.log(err || `Replaced ${path}`);
+      });
+    });
+  });
+});
+
+export function transformCode(code: string, filePath: string, removeOnDestroy = false) {
+  const sourceFile = project.createSourceFile(filePath, code, { overwrite: true });
+  replaceOldImport(sourceFile);
+
+  sourceFile.getClasses().forEach(classDeclaration => {
+    addUntilDestroyDecorator(classDeclaration);
+
+    if (removeOnDestroy) {
+      const ngOnDestroyDeclaration = classDeclaration.getMember('ngOnDestroy');
+      if (!ngOnDestroyDeclaration) return;
+
+      const ngOnDestroyIsNotEmpty = Boolean(
+        ngOnDestroyDeclaration.getDescendantStatements().length
+      );
+      if (ngOnDestroyIsNotEmpty) return;
+
+      ngOnDestroyDeclaration.remove();
+
+      removeOnDestroyImplements(classDeclaration);
+      removeOnDestroyImport(sourceFile);
+    }
+  });
+
+  return sourceFile.getFullText();
+}
+
+function replaceOldImport(sourceFile: SourceFile) {
+  const oldImport = sourceFile.getImportDeclaration('ngx-take-until-destroy');
+  oldImport &&
+    oldImport.replaceWithText(
+      `import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';`
+    );
+}
+
+function addUntilDestroyDecorator(classDeclaration: ClassDeclaration) {
+  const decorators = [
+    { name: 'UntilDestroy', arguments: [] },
+    ...(classDeclaration.getStructure().decorators || [])
+  ];
+  classDeclaration.getDecorators().forEach(d => d.remove());
+  classDeclaration.addDecorators(decorators);
+}
+
+function removeOnDestroyImplements(classDeclaration: ClassDeclaration) {
+  const onDestroyImplementClause = classDeclaration
+    .getImplements()
+    .find(impl => impl.getText() === 'OnDestroy');
+  onDestroyImplementClause && classDeclaration.removeImplements(onDestroyImplementClause);
+}
+
+function removeOnDestroyImport(sourceFile: SourceFile) {
+  const importDeclaration = sourceFile.getImportDeclaration('@angular/core');
+  if (!importDeclaration) return;
+
+  const importClause = importDeclaration.getImportClause();
+  if (!importClause) return;
+
+  const onDestroyImportSpecifier = importClause
+    .getNamedImports()
+    .find(node => node.getText() === 'OnDestroy');
+  if (!onDestroyImportSpecifier) return;
+
+  onDestroyImportSpecifier.remove();
+
+  if (!importDeclaration.getImportClause()) {
+    importDeclaration.remove();
+  }
```

Comment:
This will not be reached if `importClause` is undefined, since you do `if ! importClause return` above.

Reply:
@arturovt why? I change importClause between these checks. `check importDeclaration -> check importClause -> check onDestroy specifier -> remove importSpecifier -> check importClause and remove if it's undefined (onDestroy was only one importSpecifier)`

Reply:
You have this on the 86th line:
```ts
const importClause = importDeclaration.getImportClause();
if (!importClause) return;
```

If it's not `undefined` then all statements will be executed. On the 96th you have a check:
```ts
if (!importDeclaration.getImportClause()) { ... }
```

It will never be executed because you have the `return` statement if `importClause` is `undefined`.

Reply:
`importDeclaration` is mutable object and I check it twice. This case was covered in `Component with single OnDestroy import from "@angular/core -> should remove OnDestroy import` test.

Reply:
You have to add comments to the code then "why it's done like that". Other people that will be reading this piece of code will get confused.

Reply:
Added a corresponding comment